actionpack/lib/action_dispatch/http/uri.rb

```diff
@@ -0,0 +1,152 @@
+# frozen_string_literal: true
+
+require "active_support/core_ext/module/attribute_accessors"
+require "uri/generic"
+
+module ActionDispatch
+  module Http
+    class URI
+      mattr_accessor :tld_length, default: 1
+      attr :core_uri
+
+      delegate :scheme, :host, :port, :path, :query, :fragment, :to_s, to: :core_uri
+      delegate :scheme=, :host=, :port=, :path=, :query=, :fragment=, to: :core_uri
+
+      class << self
+        # Returns the domain part of a host given the domain level.
+        #
+        #    # Top-level domain example
+        #    extract_domain('www.example.com', 1) # => "example.com"
+        #    # Second-level domain example
+        #    extract_domain('dev.www.example.co.uk', 2) # => "example.co.uk"
+        def extract_domain(host, tld_length)
+          extract_domain_from(host, tld_length) if named_host?(host)
+        end
+
+        # Returns the subdomains of a host as an Array given the domain level.
+        #
+        #    # Top-level domain example
+        #    extract_subdomains('www.example.com', 1) # => ["www"]
+        #    # Second-level domain example
+        #    extract_subdomains('dev.www.example.co.uk', 2) # => ["dev", "www"]
+        def extract_subdomains(host, tld_length)
+          if named_host?(host)
+            extract_subdomains_from(host, tld_length)
+          else
+            []
+          end
+        end
+
+        # Returns the subdomains of a host as a String given the domain level.
+        #
+        #    # Top-level domain example
+        #    extract_subdomain('www.example.com', 1) # => "www"
+        #    # Second-level domain example
+        #    extract_subdomain('dev.www.example.co.uk', 2) # => "dev.www"
+        def extract_subdomain(host, tld_length)
+          extract_subdomains(host, tld_length).join(".")
+        end
+
+        def named_host?(host)
+          !ActionDispatch::Http::URL::IP_HOST_REGEXP.match?(host)
+        end
+
+        private
+          def extract_domain_from(host, tld_length)
+            host.split(".").last(1 + tld_length).join(".")
+          end
+
+          def extract_subdomains_from(host, tld_length)
+            parts = host ? host.split(".") : []
+            parts[0..-(tld_length + 2)]
+          end
+      end
+
+      def initialize(uri)
+        @core_uri = uri
+      end
+
+      def self.build_from_faulty_string(url_string)
```

Comment:
What is the reason you want to handle faulty strings?

Reply:
changed this ..